Add CMake integration.

marovira opened the pull request:
Adds full integration with CMake:
* Create CMake targets,
* Allow installation,
* Provide ways of setting the various configuration options at build time,
* Set up examples to be built through CMake.

This did require a few minor tweaks:
* I copied over `glad.h`, `linmath.h`, and `stb_image_write.h` from GLFW's source.
* Examples had to be updated to use the new code.

Other than that, everything was pretty straight-forward. 
I did have one question though: usually, libraries are versioned so that users can link against specific versions of the library. I'm guessing you want to keep `vkfw` unversioned? If not, then I can add the necessary CMake code to capture the version number when the library is installed.

Cvelth commented on README.md:
````diff
@@ -67,7 +90,7 @@ As all the deleters are known at compiler time and don't need to store anything
 These objects are accessible through `vkfw::Window::callbacks()` (or `accessWindowCallbacks(GLFWwindow *)` if enhanced mode is disabled) function and can be set directly, for example:
 ```c++
-my_window.callbacks().on_key = [](vkfw::Window const &, vkfw::Key key, int32_t, 
+my_window.callbacks().on_key = [](vkfw::Window const &, vkfw::Key key, int32_t,
````
Please, move formatting changes into a separate commit(s), as they have nothing to do with CMake.

Cvelth replied:
Wait, where did these changes go? Please include them as a separate commit.

Cvelth replied:
Sounds good!

Cvelth commented on README.md:
```diff
@@ -5,8 +5,31 @@ It focuses on features like type safety with enums and bitfields, STL containers
-That is to say, you still need to link against [glfw](https://github.com/glfw/glfw). These are just bindings, not a full-fledged library.
+There are a few ways in which you can integrate `vkfw` in your project:
+
+* The easiest way to integrate is to copy the header files into your source tree and then include `vkfw/vkfw.hpp`. Note that you still need to link against [glfw](https://github.com/glfw/glfw) as these are just bindings, not a full-fledged library.
```
```diff
- The easiest way to integrate is to copy the header files into your source tree and then include `vkfw/vkfw.hpp`.
+ The easiest way is to copy the header file (`vkfw/vkfw.hpp`) into your source tree and include it.
```

Cvelth commented on README.md:
````diff
@@ -5,8 +5,31 @@ It focuses on features like type safety with enums and bitfields, STL containers
+* The easiest way to integrate is to copy the header files into your source tree and then include `vkfw/vkfw.hpp`. Note that you still need to link against [glfw](https://github.com/glfw/glfw) as these are just bindings, not a full-fledged library.
+* If your project uses CMake, then you can include `vkfw` as a regular dependency.
+
+If using CMake, you'll want to link against the library like this:
+
+```cmake
+# If using as a submodule:
+add_subdirectory(<path-to-vkfw-dir>)
+
+# If using through FetchContent:
+include(FetchContent)
````
I feel like this would be a bit easier to follow if separate _options_ were split into distinct code blocks, maybe something like:

-----------------------

There are a few ways in which you can integrate `vkfw` in your project:

1. The easiest way is to copy the header file (`vkfw/vkfw.hpp`) into your source tree and include it. Note that you still need to link against [glfw](https://github.com/glfw/glfw) as these are just bindings, not a full-fledged library.

2. You can add it as a git submodule and then just
```cmake
add_subdirectory(<path-to-vkfw-dir>)
target_link_libraries(<your-project> PRIVATE vkfw)
```

3. You can also install it system-wide (see `VKFW_INSTALL` option of the provided `CMakeLists.txt), and then do
```cmake
find_package(vkfw REQUIRED)
target_link_libraries(<your-project> PRIVATE vkfw::vkfw)
```

4. You can use `FetchContent`:
```cmake
include(FetchContent)
FetchContent_Declare(
    vkfw
    GIT_REPOSITORY https://github.com/Cvelth/vkfw
)
FetchContent_MakeAvailable(vkfw)
target_link_libraries(<your-project> PRIVATE vkfw::vkfw)
```

5. You can add a port for it to your private `vcpkg` repository or otherwise adopt it in your preferred package manager, all you need is `vkfw/vkfw.hpp` header.

6. And a lot of other options...

----------------------------------

What do you think?

Cvelth commented on cmake/vkfwConfig.cmake:
```diff
@@ -0,0 +1,6 @@
+if (CMAKE_VERSION VERSION_GREATER_EQUAL 3.8.0)
+    include(CMakeFindDependencyMacro)
+    find_dependency(glfw3 REQUIRED)
+endif()
```
I wonder if different behavior under different cmake versions is a good idea...

That being said, `3.8.0` is an extremely old cmake version. Should we just add a `cmake_minimum_required` statement in here and call it a day instead?

Cvelth commented on .gitignore:
```diff
@@ -0,0 +1 @@
+compile_commands.json
```
Separate commit, please, as this is not directly related to CMake.

Cvelth replied:
Where did this go? Please include it as a separate commit!

Cvelth commented on deps/glad/glad.h:
```diff
@@ -0,0 +1,6398 @@
+/**
+ * Loader generated by glad 2.0.0-beta on Tue Aug 24 22:51:07 2021
+ *
+ * Generator: C/C++
```
1. These are dependencies for the examples and for the examples only, as such, to avoid potential confusion, let's move them to something like `example/dependencies`.

2. I'm not sure including third party headers "by copy" in a bindings repository is a good idea. Could we, please, have a submodule (or multiple) for these instead?

marovira replied:
We're going to have to bundle glad regardless as that one doesn't have a formal git repo that can be pulled from.

Since this introduces cmake, would you prefer to have cmake itself pull down the dependencies as part of the build? It would make things cleaner as a submodule would no longer be needed and it would keep the general structure you already had.

Cvelth replied:
Yep, I don't mind dynamically pulling dependencies.

It's just the committing of third party code (especially under different licenses) that I'm concerned about.

marovira replied:
Neither `linmath.h` nor `stb` have integration with CMake, so pulling them dynamically is a no-go. 
Given that `glad.h` has to be bundled in the source tree regardless of what happens, I have two proposals:
1. Have all of them moved to `example/deps` and add a note in the `README` explaining that these bundled dependencies are purely for the examples and not required by the library itself (similar to how GLFW itself does it in [theirs](https://github.com/glfw/glfw?tab=readme-ov-file#dependencies).
2. Keep `glad.h` in `examples/deps` and have `linmath` and `stb` pulled as submodules.

I think the first option is better, as the second would pull the entire `stb` library, which seems like a waste given that the examples only really need one file.

Cvelth replied:
Looking at it, I'd say there are three more options:

3. `GLFW` already uses all the same dependencies for its examples. We could just add a single submodule at `examples/dependencies/glfw` or something and pull those from there.

4. Add all the dependencies as the submodules (`linmath` and `stb` are simple, as for `glad` we (either me or you, doesn't matter) can make a new repo for it and add that as a submodule).

5. I quickly looked into `FetchContent` and it looks like it works just fine with non-cmake targets. It's not like we need any complex logic attached, just some headers. We could just add them as `URL`s.

---------------

IMO, if we can do it dynamically, we probably should. Bonus points for being able to hide this under a flag (either `VKFW_FETCH_DEPENDENCIES` (that we discussed in a different thread) or a separate flag works).

marovira replied:
How about this:
Have `VKFW_BUILD_EXAMPLES` automatically set `VKFW_FETCH_DEPENDENCIES` to true. Then I can grab the path to the dependencies from the directory itself is cloned and set that for the examples. That way there's no files bundled for the examples and everything works transparently through the existing flags.

Cvelth replied:
Right, I totally forgot we're already fetching GLFW! Good catch, that is indeed a perfect solution!

Cvelth commented on example/boing.cpp:
```diff
@@ -595,7 +596,7 @@ int main(void) {
 
     window->makeContextCurrent();
-    gladLoadGLLoader((GLADloadproc) &vkfw::getProcAddress);
+    gladLoadGL((GLADloadfunc) &vkfw::getProcAddress);
```
These are my bad 😅.

That being said, since they are not related to CMake, please move them into a separate commit.

Cvelth commented on example/CMakeLists.txt:
```diff
@@ -0,0 +1,52 @@
+
+set(LINK_LIBS vkfw glfw Vulkan::Vulkan)
+
+add_executable(boing ${VKFW_EXAMPLE_ROOT}/boing.cpp)
+target_link_libraries(boing PRIVATE ${LINK_LIBS})
+target_include_directories(boing PRIVATE ${VKFW_DEPS_ROOT})
```
I think we should take a bit of inspiration from the corresponding glfw targets for these: https://github.com/glfw/glfw/blob/e7ea71be039836da3a98cea55ae5569cb5eb885c/examples/CMakeLists.txt

In particular for MSVC and Apple specific flags.

Cvelth commented on CMakeLists.txt:
```diff
@@ -0,0 +1,136 @@
+cmake_minimum_required(VERSION 3.20)
+project(vkfw LANGUAGES CXX C)
```
I wonder if there's a point in including `C` here 🤔.

Cvelth commented on CMakeLists.txt:
```diff
@@ -0,0 +1,136 @@
+project(vkfw LANGUAGES CXX C)
+
+# Build options.
+option(VKFW_DEV_BUILD "Build VKFW for development" OFF)
```
This seems to be used exclusively to auto-fetch GLFW, maybe the name could reflect that a bit more clearly?

marovira replied:
I left it open in the event that you wanted it to include additional behaviour. For example, dev builds could enable building the examples by default.
In the interest of future proofing, I could change the name to `VKFW_FETCH_DEPENDENCIES`. What do you think?

Cvelth replied:
> I could change the name to `VKFW_FETCH_DEPENDENCIES`

Sounds perfect!

Cvelth commented on CMakeLists.txt:
```diff
@@ -0,0 +1,136 @@
+option(VKFW_INSTALL "Create install target" OFF)
+option(VKFW_BUILD_EXAMPLES "Build examples" OFF)
+
+# Configuration options
```
It's probably worth adding something like "see `README.md` for details" to this comment.

Cvelth commented on CMakeLists.txt:
```diff
@@ -0,0 +1,136 @@
+set(VKFW_EXAMPLE_ROOT ${VKFW_SOURCE_DIR}/example)
+set(VKFW_DEPS_ROOT ${VKFW_SOURCE_DIR}/deps)
+
+include(FetchContent)
```
Please, move this inside the `if` using it.

Cvelth commented on CMakeLists.txt:
```diff
@@ -0,0 +1,136 @@
+target_link_libraries(vkfw INTERFACE glfw)
+
+# Set options
+if (VKFW_DISABLE_ENHANCED_MODE)
+    target_compile_definitions(vkfw INTERFACE VKFW_DISABLE_ENHANCED_MODE)
+endif()
+
+if (VKFW_NO_EXCEPTIONS)
+    target_compile_definitions(vkfw INTERFACE VKFW_NO_EXCEPTIONS)
+endif()
```
Please introduce a [function](https://cmake.org/cmake/help/v3.20/command/function.html) for this to reduce repetition a bit.

marovira replied:
So: the name in the if and the name being passed to `target_compile_definitions` are two different things. The first one is a boolean value, while the second is being interpreted as a string to pass in to the compiler flags. Unfortunately the only way to merge them would be for CMake to have some form of reflection for me to grab the name of the variable itself and pass it it, which CMake does not have.
Vulkan Hpp has this same problem actually and they do the same thing I'm doing here.

I could make a function, but all it would remove is the if blocks into something like this:

```cmake
add_define(
    TARGET boing
    FLAG VKFW_... 
    DEFINE VKFW_... 
) 

```

Which I can do but wouldn't really solve the name repetition.

Cvelth replied:
What do you mean? It's CMake we're talking about, like the most scripty-reflectiony-dynamicy language ever. For better or worse.

Either way, what I have in mind is something like this: https://godbolt.org/z/vvsv65Kbb (see the execution output).

marovira replied:
That's not going to work. The value assigned for boolean `option(...)` is either `ON` or `OFF`, therefore adding `target_compile_definitions(<target> INTERFACE <flag>)` will result in `ON` being added to the pre-processor, which is not what we want.

Cvelth replied:
Look at the `main.cpp`, the example above works.

If you are still hesitating, try adding
```cmake
message(${ARGUMENT}) # the name
message(${${ARGUMENT}}) # actual value (on or off)
```
lines to the script above.

marovira replied:
Ignore me. I missed the nested expansion in the scope of the function. I'll update the code accordingly.

Cvelth reviewed (approved):
Aside from a couple of minor comments, this looks really good!

Huge thank you for taking the time to do this, @marovira!

Cvelth commented:
> I'm guessing you want to keep `vkfw` unversioned?

`vkfw` is ad-hoc versioned in a sense that each version is tied to the specific version of `glfw`.

There's an extremely basic assertion [here](https://github.com/Cvelth/vkfw/blob/434ae60454df436b8aefccfe11ef67605f1e750b/include/vkfw/vkfw.hpp#L116-L118) and it does the job because new `glfw` versions only release so often. That being said, I'm not opposed to something a bit more structured.

marovira commented:
Before I start fixing things: do you prefer fixes being force pushed to the main commit or having them as separate commits?

Cvelth commented:
> Before I start fixing things: do you prefer fixes being force pushed to the main commit or having them as separate commits?

Let's keep the git history as clean as possible - that is, feel free to force push it however much you want.

marovira commented:
Here's the new set of changes. I'm going to be pushing the updates to the examples in a separate commit, but I wanted to outline what I changed first:
1. I added the fixes to the README.
2. I removed the `if` in `vkfwConfig.cmake` as it is actually not needed according to the newer docs.
3. I've refactored the code for adding the flags to use functions.
4. The examples also use a new function to simplify the code and ensure the flags from GLFW are added correctly.
5. The biggest trick is in the way I'm managing the inclusion of GLFW's source. I think the documentation I've added covers all the details, but let me know if you want me to add anything

If these changes are good, then I'll push the commit with the fixes for the examples.

marovira commented:
This should be it for this PR. Examples now build and everything is wired in!

Cvelth commented on CMakeLists.txt:
```diff
@@ -0,0 +1,108 @@
+project(vkfw LANGUAGES CXX)
+
+# Build options.
+option(VKFW_FETCH_DEPENDENCIES "Build VKFW for development" OFF)
```
The comment here is outdated :)

Cvelth commented:
Looks good! Huge thank you for making this happen, @marovira!
